018-cli-2.0/018-cli-2.0.md

```diff
@@ -0,0 +1,312 @@
+---
+ 
+Until now, the CLI was based off of our Server Side Swagger Spec. While this was a good start, it limits our functionality to a Server SDK. Ideally we want the CLI to be an alternative to the Console SDK ( aka the Appwrite Dashboard ). Here are some changes that we will need to make to adhere to the server spec.
+
+After the implementation of this RFC, the CLI will behave more like the Appwrite Console rather than a Server SDK. This accounts for a new form of Authentication that is similar to the way we handle logins in the Appwrite console. We use a cookie.
+
+We will introduce a new command `appwrite login` that will work in one of either ways.
+
+* `appwrite login` makes a request to the existing `accounts.createSession` endpoint, captures the returned cookie and uses this cookie in all the subsequent requests. 
+
+    Whenever we wish to access console specific features, we can authenticate those requests using the cookie and continue to use the API Key based authentication when using the CI.
+
+    ```sh
+    $ appwrite login
+    Email:
+    Password:
+    ```
+
+    In CI environments, the authentication will continue to take place using the API key on a per project basis until we develop a new token / key based mechanism for console authentication. 
+
+    ```sh
+    appwrite client --setKey="" --setEndpoint="" ...
+    ```
+
+* Second method relies on a browser based auth.
+
+  ```sh
+  $ appwrite login
+
+  # Browser based authentication continues. After a successful login, the returned token and user ID are stored in a hidden preferences file.
+  ```
+
+  1. The command first finds an available local port. 
+  2. Construct a redirect URL using the available port `https://localhost:1234`
+  3. Spin up a local http server listening to on the available port ( 1234 in this case )
+  4. Opens the browser with a request to either the local appwrite server or appwrite cloud `/login?callback=https://localhost:1234` with the redirect URL as one of the query parameters.
+
+  ![CLI Auth](cli_auth.png)
+
+  This approach may seem unnecessarily complex now since we only allow email password based logins in the console. But as we move towards supporting OAuth logins in the console, this is the approach that needs to be followed. [Firebase CLI](https://github.com/firebase/firebase-tools/blob/0b0459bfde3fabf41223bb6d5d39a5cf325f1010/src/auth.ts#L476) for reference. 
+
+
+## Easier Deployment of cloud functions
+--- 
+
+### 🟢 Constrain deployments from current directory.
+
+We start by removing the `--code` parameter in the `createTag` command and constrain function deployments only from the current directory. 
+
+The refactored command will look like this 
+
+```sh
+appwrite functions createTag --functionId=[ID] --command=[COMMAND]
+```
+
+This corresponding appwrite issue will also need to be tackled as we implement this.
+https://github.com/appwrite/appwrite/issues/1316
+
+### 🟢 The `.appwrite.json` File
```

Comment:
Is there any benefit from a hidden file? Should we just use `appwrite.json`?

Reply:
I made it hidden under the assumption that we will need to store secrets in there. But since we planned to store the secrets in another file, this can be changed to  `appwrite.json`